contracts/mirror_staking/src/staking.rs

```diff
@@ -148,6 +152,134 @@ pub fn decrease_short_token<S: Storage, A: Api, Q: Querier>(
+                msg: to_binary(&Cw20HandleMsg::TransferFrom {
+                    owner: env.message.sender.clone(),
+                    recipient: env.contract.address.clone(),
+                    amount: token_amount,
+                })?,
+                send: vec![],
+            }),
+            CosmosMsg::Wasm(WasmMsg::Execute {
+                contract_addr: token_addr.clone(),
+                msg: to_binary(&Cw20HandleMsg::IncreaseAllowance {
+                    spender: terraswap_pair.contract_addr.clone(),
+                    amount: token_amount,
+                    expires: None,
+                })?,
+                send: vec![],
+            }),
+            CosmosMsg::Wasm(WasmMsg::Execute {
+                contract_addr: terraswap_pair.contract_addr,
+                msg: to_binary(&PairHandleMsg::ProvideLiquidity {
+                    assets: [
+                        Asset {
+                            amount: (native_asset.amount.clone() - tax_amount)?,
+                            info: native_asset.info.clone(),
+                        },
+                        Asset {
+                            amount: token_amount,
+                            info: AssetInfo::Token {
+                                contract_addr: token_addr.clone(),
+                            },
+                        },
+                    ],
+                    slippage_tolerance,
+                })?,
+                send: vec![native_asset.deduct_tax(deps)?],
+            }),
+            CosmosMsg::Wasm(WasmMsg::Execute {
+                contract_addr: env.contract.address,
+                msg: to_binary(&HandleMsg::AutoStakeHook {
+                    asset_token: token_addr.clone(),
+                    staking_token: terraswap_pair.liquidity_token,
+                    staker_addr: env.message.sender,
+                })?,
+                send: vec![],
+            }),
+        ],
+        log: vec![
+            log("action", "auto_stake"),
+            log("asset_token", token_addr.to_string()),
+            log("tax_amount", tax_amount.to_string()),
+        ],
+        data: None,
+    })
+}
+
+pub fn auto_stake_hook<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    asset_token: HumanAddr,
+    staking_token: HumanAddr,
+    staker_addr: HumanAddr,
+) -> HandleResult {
+    // only can be called by itself
+    if env.message.sender != env.contract.address {
+        return Err(StdError::unauthorized());
+    }
+
+    // assert staking token
+    let pool_info: PoolInfo =
+        read_pool_info(&deps.storage, &deps.api.canonical_address(&asset_token)?)?;
+
+    if pool_info.staking_token != deps.api.canonical_address(&staking_token)? {
+        return Err(StdError::generic_err("Invalid staking token"));
+    }
+
+    // stake all lp tokens received
+    let staking_token_amount = query_token_balance(&deps, &staking_token, &env.contract.address)?;
```

Comment:
We need to pass previous balance and subtract that from the current balance to get actual staking amount

Reply:
Fixed it